// @azure-tools/cadl-rpaas/dist/src/commonTypes.js
import { $useRef } from "@azure-tools/cadl-autorest";
import { reportDiagnostic } from "./lib.js";
export function getArmTypesPath(program) {
    return (program.getOption("arm-types-path") ||
        "../../../../../common-types/resource-management/v2/types.json");
}
export function $armCommonDefinition(program, entity, definitionName) {
    if (entity.kind !== "Model") {
        reportDiagnostic(program, {
            code: "decorator-wrong-type",
            messageId: "armCommonDefinition",
            target: entity,
        });
        return;
    }
    // Use the name of the model type if not specified
    if (!definitionName) {
        definitionName = entity.name;
    }
    $useRef(program, entity, `${getArmTypesPath(program)}#/definitions/${definitionName}`);
}
export function $armCommonParameter(program, entity, parameterName) {
    if (entity.kind !== "ModelProperty") {
        reportDiagnostic(program, {
            code: "decorator-wrong-type",
            messageId: "armCommonParameter",
            target: entity,
        });
        return;
    }
    // Use the name of the model type if not specified
    if (!parameterName) {
        parameterName = entity.name;
    }
    $useRef(program, entity, `${getArmTypesPath(program)}#/parameters/${parameterName}`);
}
